### ISSUE TITLE
Using NextToken in Orders API (Python)

### ISSUE THREAD

**Issue by jerednel** [amzn/selling-partner-api-docs#1777]
The documentation says to just pass the NextToken value, if present, to the same API call in order to paginate to the next list of orders.

I can get the first 100 orders just fine but when I try to do as the documentation says I get an error saying `The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details`

```python 
def get_orders(...)
....
...
        ordersResponse = requests.get("https://sellingpartnerapi-eu.amazon.com/orders/v0/orders", 
                                    params=params, headers=headers, auth=auth).json()
        orders = ordersResponse['payload']['Orders']
            
        while ordersResponse['payload'].get('NextToken'):
            print("Paginating")
            params['NextToken'] = ordersResponse['payload']['NextToken']
            ordersResponse = requests.get("https://sellingpartnerapi-eu.amazon.com/orders/v0/orders", 
                                        params=params, headers=headers, auth=auth)
            print(ordersResponse.text)
            orders.extend(ordersResponse['payload']['Orders'])
       
```

I cannot find what I am supposed to do in order to make the calls with NextToken work.  This cannot be as simple as it says here https://github.com/amzn/selling-partner-api-docs/blob/main/references/orders-api/ordersV0.md#description "If NextToken is present, that will be used to retrieve the orders instead of other criteria.".  It is present for sure.  Does it need to be present in a different way?

**Comment by stefnats** [comment]
When you change the params, the headers also have to be recalculated.
The headers contain the signature which is calculated on several parameters, including the params. So when one parameter changes, the Signature is rendered invalid.

**Comment by jerednel** [comment]
Thank you for that. Should I be including the NextToken in the headers as "NextToken"?  I am using a structure similar to below where I am setting an API object, authorizing it, then using that to get orders.  I am still running into the same issue though.  When I try to pass the token to the get_orders function and then have it set in both the headers AND the params it still gives me the same 403 error.

api class
```python
def auth(self):
        ...other_stuff..here...

        access_token = lwa.json()['access_token']
        # Create headers to add to SP API request. Headers should include: content_type and "x-amz-access-token"
        self.headers = {'content-type': 'application/json','Accept': 'application/json','x-amz-access-token':access_token}

        auth = AWS4Auth(AccessKeyId, SecretAccessKey, 'eu-west-1', 'execute-api',
                        session_token=SessionToken)
        self.auth = auth

def get_orders(self, marketplaceId, next_token=None):
        headers = self.headers
        auth = self.auth

        createdAfter = '2021-09-15T00:00:00.000000'
        createdBefore = '2021-10-30T00:00:00.000000'

        params = {'CreatedAfter': createdAfter, 'CreatedBefore': createdBefore, 'MarketplaceIds': [marketplaceId], 'OrderStatuses': ['Shipped']}
        if next_token is not None:
            params.update({'NextToken':next_token})
            headers.update({'NextToken':next_token})

        ordersResponse = requests.get("https://sellingpartnerapi-eu.amazon.com/orders/v0/orders", 
                                    params=params, headers=headers, auth=auth)
```

main
```python

api_client = api.APIClient(aws_access_key=os.environ.get('AWS_ACCESS_KEY'),
                               aws_secret_key=os.environ.get('AWS_SECRET_KEY'),
                            aws_account_id=os.environ.get('AWS_ACCOUNT_ID'),
                            client_id=os.environ.get('CLIENT_ID'),
                            client_secret=os.environ.get('CLIENT_SECRET'),
                               refresh_token=os.environ.get('REFRESH_TOKEN'))

api_client.auth()

ordersresponse = api_client.get_orders(marketplaceid, since)
orders.append(ordersresponse['payload']['Orders'])
while ordersresponse['payload'].get('NextToken'):
    next_token = ordersresponse['payload']['NextToken']
    ordersresponse = api_client.get_orders(marketplaceid, since, next_token)
```

**Comment by RogerGatco** [comment]
In your auth Method
Try changing 
` self.headers = {'content-type': 'application/json','Accept': 'application/json','x-amz-access-token':access_token}`

to use application/x-www-form-urlencoded instead

` self.headers = {'content-type': 'application/x-www-form-urlencoded,'Accept': 'application/json','x-amz-access-token':access_token}`

**Comment by kpconnell** [comment]
@jerednel you are reusing params for two totally different api calls. bad bad coding choice.  repopulate a new param object using the same way you did for the initial call, and after setting the next token (which is the only parameter)

**Comment by jerednel** [comment]
I appreciate the feedback but even if I create a new params object containing just the NextToken I still get a invalid signature error.  I promise I am trying to spend time to understand why I could be running into this but it's just not happening.  

In case I am omitting something important in my explanations, here is a beginning-to-end version of the application that is giving me the same error.  I've also tried urllib.parse.quoteing/quote_plusing the next_token and that doesn't seem to get me any further.

```python
import requests
import boto3
from requests_aws4auth.aws4auth import AWS4Auth
import os
from dotenv import load_dotenv
load_dotenv()

ACCESS_ID=os.environ.get('AWS_ACCESS_KEY')
ACCESS_KEY=os.environ.get('AWS_SECRET_KEY')
ACCOUNT_ID=os.environ.get('AWS_ACCOUNT_ID')
CLIENT_ID=os.environ.get('CLIENT_ID')
CLIENT_SECRET=os.environ.get('CLIENT_SECRET')
REFRESH_TOKEN=os.environ.get('REFRESH_TOKEN')

def main():
    orders = get_orders()
    print(orders)
    
def get_orders():
    url = "https://sellingpartnerapi-eu.amazon.com/orders/v0/orders/"
    AccessKeyId, SecretAccessKey, SessionToken, AccessToken = get_session_data()
    headers = {
            'x-amz-access-token': AccessToken, 
            'content-type': 'application/x-www-form-urlencoded', 
            'Accept':'application/json'
        }
    auth = AWS4Auth(AccessKeyId, SecretAccessKey, 'eu-west-1', 'execute-api', session_token=SessionToken)

    params = {
        'MarketplaceIds': 'A13V1IB3VIYZZH',
        'CreatedAfter': '2021-11-01T00:00:00',
        'CreatedBefore':  '2021-11-28T00:00:00',
        'OrderStatuses': 'Shipped',
    }
    r = requests.get(url, params=params, auth=auth, headers=headers)
    response = r.json()['payload']
    orders = response['Orders']
    orders_list = list()
    orders_list.extend(orders)
    while True:
        if response.get('NextToken'):
            next_token = response['NextToken']
            orders_list.extend(get_paginated_orders(next_token=next_token, auth=auth, access_token=AccessToken))
        else:
            break

    return orders_list

def get_paginated_orders(**kwargs):
    url = f"https://sellingpartnerapi-eu.amazon.com/orders/v0/orders/"

    headers = {
            'x-amz-access-token': kwargs['access_token'], 
            'content-type': 'application/x-www-form-urlencoded', 
            'Accept':'application/json',
        }
    params = {
        'NextToken': kwargs['next_token']
    }

    r = requests.get(url, auth=kwargs['auth'], headers=headers, params=params)

    print(f"Show me the error...\n{r.text}")
    response = r.json()['payload']
    orders = response['Orders']
    return orders
    
def get_session_data():
    session = boto3.session.Session(aws_access_key_id=ACCESS_ID,
                                    aws_secret_access_key=ACCESS_KEY
                                    )
    client = session.client('sts')
    iamrole = 'arn:aws:iam::'+ACCOUNT_ID+':role/SPAPIIAMRole'

    response = client.assume_role(
        RoleArn = iamrole,
        RoleSessionName = 'SPAPIRoleSession'
    )
    AccessKeyId = response['Credentials']['AccessKeyId']
    SecretAccessKey = response['Credentials']['SecretAccessKey']
    SessionToken = response['Credentials']['SessionToken']

    payload = {'grant_type':'refresh_token','client_id':CLIENT_ID,'client_secret':CLIENT_SECRET,'refresh_token':REFRESH_TOKEN}

    lwa = requests.post("https://api.amazon.com/auth/o2/token", data=payload)

    AccessToken = lwa.json()['access_token']
    return AccessKeyId, SecretAccessKey, SessionToken, AccessToken


if __name__=='__main__':
    main()
```

**Comment by RogerGatco** [comment]
I think the issue lies in the get_paginated_orders function

you are reusing your signed signature from the initial call but it needs to be recalculated and signed with the new parameters and headers.

You can try adding the following code in the get_paginated_orders  function right before you make the get call.


kwargs['auth'] = AWS4Auth(AccessKeyId, SecretAccessKey, 'eu-west-1', 'execute-api', session_token=SessionToken)

**Comment by RogerGatco** [comment]
This is how ours setup (Using C# .NET for getPurchaseOrders). I'm assuming it's similar to what you are trying to do. I have gotten that same error message before and resolved it like this.

```csharp

// Amazon SignatureHelper 
private AuthHelper signatureHelper;
// About to call with nextToken
while (amzCall.payload.pagination != null)
{
	//1 - Building request
         RestClient restClient = new RestClient("https://sellingpartnerapi-na.amazon.com");
	 IRestRequest restRequest = new RestRequest("/vendor/orders/v1/purchaseOrdersStatus?createdAfter=blahblah&nextToken=blahblahblah" , Method.GET);
	
       //2 -  Adding token header
	restRequest.AddHeader("x-amz-access-token", "access token here");
	
	//3 - Signing it
	var request2 = signatureHelper.SignRequest(restRequest, restClient, "application/x-www-form-urlencoded");
	
	//4 -  Making GET call
	var resultNextPage = restClient.Execute(request2);

}
```

Hope this helps.

**Comment by radug-cpc** [comment]
@jerednel , did you manage to find a fix to your problem?

**Comment by iozbek** [comment]
I had the same issue with the getOrders calls with NextToken parameter. It works for the first call without NextToken and retrieves 100 orders but then with the NextToken same call gives an error as "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details."

We implemented the signing process by following the documentation https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html. 

In the documentation it says "Double-encode any equals ( = ) characters in parameter values." I thought it might be the problem when we double encode an equal character in a token parameter. So I encoded equal character once for the NextToken value and it worked! 

Hence, either the documentation or the implementation should be updated.
